src/dftbp/math/wignerseitz.F90

```diff
@@ -0,0 +1,216 @@
+  !! PySCF: the Python-based simulations of chemistry framework.
+  !! WIREs Comput Mol Sci, 8: e1340.
+  !! DOI: 10.1002/wcms.1340
+  subroutine generateWignerSeitzGrid(nKpt, latVecs, wsVectors, wsSearchSize, distTol)
+
+    !> Supercell folding coefficients (diagonal elements, i.e. MP-type grid)
+    integer, intent(in) :: nKpt(:)
+
+    !> Real-space lattice vectors of periodic geometry
+    real(dp), intent(in) :: latVecs(:,:)
+
+    !> Lattice vectors inside/on the surface of the WS-supercell in fractional coordinates
+    integer, intent(out), allocatable :: wsVectors(:,:)
+
+    !> Optional index range to search for primitive unit cells inside the WS-supercell
+    integer, intent(in), optional :: wsSearchSize(:)
+
+    !> Optional tolerance for comparing real-space distances
+    real(dp), intent(in), optional :: distTol
+
+    !> Actual tolerance for comparing real-space distances
+    real(dp) :: distTol_
+
+    !! Actual index range to search for primitive unit cells inside the WS-supercell
+    integer :: wsSearchSize_(3)
+
+    !> Lattice vectors inside/on the surface of the WS-supercell in Cartesian coordinates
+    real(dp), allocatable :: wsVectorsCart(:,:)
+
+    !! Work buffer of real-space lengths in Cartesian coordinates
+    real(dp), allocatable :: wsLengths(:)
+
+    !! Index array for sorting operations
+    integer, allocatable :: ind(:)
+
+    !! Temporary distance index that indicates the central cell
+    integer :: iCenter
+
+    !! Auxiliary variables
+    integer :: n1, n2, n3, i1, i2, i3, ii, jj, diff(3), tmp3Int(3)
+    real(dp) :: dotProducts(3, 3), diffMat(3, 3)
+
+    !! Temporary storage for distances
+    real(dp), allocatable :: dist(:)
+
+    @:ASSERT(all(nKpt > 0))
+    @:ASSERT(size(latVecs, dim=2) == 3)
+
+    if (present(wsSearchSize)) then
+      @:ASSERT(size(wsSearchSize, dim=1) == 3)
+      wsSearchSize_(:) = wsSearchSize
+    else
+      wsSearchSize_(:) = 2
+    end if
+
+    if (present(distTol)) then
+      @:ASSERT(distTol > 0.0_dp)
+      distTol_ = distTol
+    else
+      distTol_ = 1.0e-08_dp
+    end if
+
+    ! Cache repeatedly used dot-products
+    do jj = 1, size(latVecs, dim=2)
+      do ii = 1, size(latVecs, dim=2)
+        dotProducts(ii, jj) = dot_product(latVecs(:, ii), latVecs(:, jj))
+      end do
+    end do
+
+    ! We brute-force search over a "large enough" supercell of the primitive Wigner-Seitz supercell
+    ! to find all primitive unit cells inside/on the surface of the primitive WS-supercell.
+    do n1 = -wsSearchSize_(1) * nKpt(1), wsSearchSize_(1) * nKpt(1)
+      do n2 = -wsSearchSize_(2) * nKpt(2), wsSearchSize_(2) * nKpt(2)
+        do n3 = -wsSearchSize_(3) * nKpt(3), wsSearchSize_(3) * nKpt(3)
+          if (allocated(dist)) deallocate(dist)
+          do i1 = -wsSearchSize_(1), wsSearchSize_(1)
+            do i2 = -wsSearchSize_(2), wsSearchSize_(2)
+              do i3 = -wsSearchSize_(3), wsSearchSize_(3)
+                diff(1) = n1 - i1 * nKpt(1)
```

Comment:
Compiler optimisation is probably moving this already, but can evaluate after line 112

Reply:
True, consider it done.